### PR TITLE
[BUG-3581] [Code Style] HttpStatus & Headers should not be imported from `org.apache.hc.core5.http`

### PR DESCRIPTION
**Pull request by prabhask5**
### Description

Accounts for refactor of Header and HttpStatus classes from hc.cor5.http to just http by adding checkstyle test, and manually refactoring all current occurrences of the illegal import. Additionally updated spotlessApply to make the refactor automatic on future cases of illegal import.

* Category (Enhancement, New feature, Bug fix, Test fix, Refactoring, Maintenance, Documentation)

Bug fix/refactor

* Why these changes are required?

Refer to #3581 

* What is the old behavior before changes and new behavior after changes?

Ideally nothing.

### Issues Resolved
- #3581 

Is this a backport? If so, please add backport PR # and/or commits #
No.

### Testing
Tests have been modified and should work before merging.

### Check List
- [x] New functionality includes testing
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by prabhask5** [comment]
@cwperks @scrawfor99 @DarshitChanpura @peternied Through updating the illegal import `org.apache.hc.core5.http.Header`, to `org.apache.http.Header` I'm getting a lot of type compilation errors that try unsuccessfully to cast `org.apache.hc.core5.http.Header` to `org.apache.http.Header`. Here is an example of one such error: /Users/prabhask/Documents/Internships/opensearch-security/src/test/java/org/opensearch/security/test/helper/rest/RestHelper.java:238: error: incompatible types: org.apache.http.Header cannot be converted to org.apache.hc.core5.http.Header

To solve this problem I can either manually add a cast to `org.apache.hc.core5.http.Header` whenever this error appears or change the imports of all the dependent classes to also exclude core5 or client5, both of which seem like bad solutions. Is there another way to solve this problem?

**Comment by reta** [comment]
> @cwperks @scrawfor99 @DarshitChanpura @peternied Through updating the illegal import `org.apache.hc.core5.http.Header`,

@prabhask5 why  `org.apache.hc.core5.http.Header` is illegal? The `main` branch uses Apache HttpClient 5 and should import these packages, the `2.x` uses Apache HttpClient 4.x and should import `org.apache.http.Header`, there should be no attempts to do any kind of type casting.

**Comment by prabhask5** [comment]
> > @cwperks @scrawfor99 @DarshitChanpura @peternied Through updating the illegal import `org.apache.hc.core5.http.Header`,
> 
> @prabhask5 why `org.apache.hc.core5.http.Header` is illegal? The `main` branch uses Apache HttpClient 5 and should import these packages, the `2.x` uses Apache HttpClient 4.x and should import `org.apache.http.Header`, there should be no attempts to do any kind of type casting.

I'm referencing #3581 , if its not illegal I might be misunderstanding what the issue is saying.

**Comment by reta** [comment]
> I'm referencing #3581 , if its not illegal I might be misunderstanding what the issue is saying.

Me neither, the `security` plugin uses `RestClient` which directly exports some of the Apache HttpClient 5 or 4.x APIs (depending on the branch).

**Comment by prabhask5** [comment]
> @prabhask5 thanks for being proactive on this.
> 
> Similar to @reta:
> 
> I am not sure what the original issue was referencing. I don't think that it will be possible to standardize the versions without _significant_ refactoring. You will need to swap to the v4 equivalents for everything in main.
> 
> Eventually we will move to 3.0 and likely swap completely to v5 when we do.
> 
> I don't think it makes sense to spend too much time converting the versions back and forth.

So is this issue irrelevant then? @peternied Do you have some insight on this?

**Comment by peternied** [comment]
Illegal has weighty connotations - I'd prefer to think of direct imports of `...hc.core5...` as _not advisable_.  The rational is that it makes our great auto-backporting tool not work for nearly any case where we are authoring test cases (which I hope is often!).  There is more context in the issue [1].

@reta What do you think about this rational? After we release 3.0.0, we will still be supporting 2.x builds for a least a year, anything that helps make the backport process result in less merge conflicts is a solid win to me.  

- [1] https://github.com/opensearch-project/security/issues/3581

**Comment by peternied** [comment]
@prabhask5 Looks like there are many merge conflicts that you'll need to resolve - I'd suggest a `git merge origin/main` YMMV

**Comment by prabhask5** [comment]
> @prabhask5 Looks like there are many merge conflicts that you'll need to resolve - I'd suggest a `git merge origin/main` YMMV

@peternied Sounds good. How do you recommend fixing the errors I'm getting? I have them copied below for convenience.

Through updating the illegal import org.apache.hc.core5.http.Header, to org.apache.http.Header I'm getting a lot of type compilation errors that try unsuccessfully to cast org.apache.hc.core5.http.Header to org.apache.http.Header. Here is an example of one such error: /Users/prabhask/Documents/Internships/opensearch-security/src/test/java/org/opensearch/security/test/helper/rest/RestHelper.java:238: error: incompatible types: org.apache.http.Header cannot be converted to org.apache.hc.core5.http.Header

To solve this problem I can either manually add a cast to org.apache.hc.core5.http.Header whenever this error appears or change the imports of all the dependent classes to also exclude core5 or client5, both of which seem like bad solutions. Is there another way to solve this problem?

**Comment by peternied** [comment]
`org.apache.hc.core5.http.Header` and `org.apache.http.Header` are different implementations, they won't translate between one another.  There are a couple of options to handle this:
1) Don't attempt to convert these objects.  While it isn't idea,l if the majority of the time we are using status codes and header constants ARE converted, that solves the majority problem.
2) Create a shared interface / use another object at object declartions.  Headers aren't very complex, it would be relatively easy to create or use another object that represented this key value pairing, then convert it in the RestHelper where it needs to be type specific.
3) An even better solution some invents.

**Comment by prabhask5** [comment]
> `org.apache.hc.core5.http.Header` and `org.apache.http.Header` are different implementations, they won't translate between one another. There are a couple of options to handle this:
> 
> 1. Don't attempt to convert these objects.  While it isn't idea,l if the majority of the time we are using status codes and header constants ARE converted, that solves the majority problem.
> 2. Create a shared interface / use another object at object declartions.  Headers aren't very complex, it would be relatively easy to create or use another object that represented this key value pairing, then convert it in the RestHelper where it needs to be type specific.
> 3. An even better solution some invents.

@peternied I'm pretty confused on these solutions. The majority of the errors I'm getting are either functions where the new Header is not an applicable type for the parameter, but I can't edit these functions since they are in the Apache codebase. How should I fix these errors? The only solution I'm seeing is to manually cast to the new type of Header, which again I don't think is advisable. A more specific solution would help a lot, ty!

**Comment by peternied** [comment]
@prabhask5 I might recommend going with the first option, remove changes for `Headers` for this PR.  After the status codes have been aligned take a single file and see what it takes to get it working with a draft PR.

This change is already massive and providing feedback is non-trivial if you are running into build/runtime issues.

**Comment by prabhask5** [comment]
> @prabhask5 I might recommend going with the first option, remove changes for `Headers` for this PR. After the status codes have been aligned take a single file and see what it takes to get it working with a draft PR.
> 
> This change is already massive and providing feedback is non-trivial if you are running into build/runtime issues.

Done! Feel free to unblock tests.

File: src/integrationTest/java/org/opensearch/security/http/JwtAuthenticationTests.java [hunk 1]
```diff
@@ -16,6 +16,8 @@
 import java.util.Map;
 
 import com.carrotsearch.randomizedtesting.annotations.ThreadLeakScope;
+import io.jsonwebtoken.SignatureAlgorithm;
+import io.jsonwebtoken.security.Keys;
 import org.apache.hc.core5.http.Header;
 import org.apache.hc.core5.http.message.BasicHeader;
```
Comment (DarshitChanpura) [inline src/integrationTest/java/org/opensearch/security/http/JwtAuthenticationTests.java:22]: this seems to be unrelated. The original import for hc5 still remains as is

File: src/integrationTest/java/org/opensearch/security/http/OnBehalfOfJwtAuthenticationTest.java [hunk 2]
```diff
@@ -21,8 +21,12 @@
 import javax.crypto.SecretKey;
 
 import com.carrotsearch.randomizedtesting.annotations.ThreadLeakScope;
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.security.Keys;
+
 import org.apache.hc.core5.http.Header;
```
Comment (DarshitChanpura) [inline src/integrationTest/java/org/opensearch/security/http/OnBehalfOfJwtAuthenticationTest.java:28]: why only 1 of these 2 changed?

File: src/integrationTest/java/org/opensearch/security/rest/CompressionTests.java [hunk 3]
```diff
@@ -20,7 +20,7 @@
 import com.carrotsearch.randomizedtesting.annotations.ThreadLeakScope;
 import org.apache.hc.client5.http.classic.methods.HttpPost;
 import org.apache.hc.core5.http.ContentType;
-import org.apache.hc.core5.http.HttpStatus;
+import org.apache.http.HttpStatus;
```
Comment (DarshitChanpura) [inline src/integrationTest/java/org/opensearch/security/rest/CompressionTests.java:23]: should this be a mixed usage between 4 & 5?

File: src/integrationTest/java/org/opensearch/security/rest/WhoAmITests.java [hunk 4]
```diff
@@ -22,7 +22,8 @@
 import java.util.stream.Collectors;
 
 import com.carrotsearch.randomizedtesting.annotations.ThreadLeakScope;
-import org.apache.hc.core5.http.HttpStatus;
+import joptsimple.internal.Strings;
```
Comment (DarshitChanpura) [inline src/integrationTest/java/org/opensearch/security/rest/WhoAmITests.java:25]: what is this import? is it being used?

File: src/test/java/org/opensearch/security/system_indices/AbstractSystemIndicesTests.java [hunk 5]
```diff
@@ -15,6 +15,7 @@
 import java.util.List;
 
 import org.apache.hc.core5.http.Header;
+
```
Comment (DarshitChanpura) [inline src/test/java/org/opensearch/security/system_indices/AbstractSystemIndicesTests.java:18]: nit: seems like this import was not replaced

File: src/test/java/org/opensearch/security/system_indices/SystemIndexDisabledTests.java [hunk 6]
```diff
@@ -11,10 +11,8 @@
 
 package org.opensearch.security.system_indices;
 
-import java.io.IOException;
-
 import org.apache.hc.core5.http.Header;
```
Comment (DarshitChanpura) [inline src/test/java/org/opensearch/security/system_indices/SystemIndexDisabledTests.java:14]: should this be org.apache.http.HttpStatus

**Review by DarshitChanpura — COMMENTED** [review]
I see mixed usages of 4 & 5. Also some classes still import Header from 5 vs 4, is that expected?
@reta thoughts?

**Comment by prabhask5** [comment]
Closed PR due to git issues new PR here: https://github.com/opensearch-project/security/pull/3741
